exporter/loadbalancingexporter/resolver_k8s_test.go

```diff
@@ -0,0 +1,169 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package loadbalancingexporter
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes/fake"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func TestK8sResolve(t *testing.T) {
+	// prepare
+	service, defaultNs, ports := "lb", "default", []int32{8080, 9090}
+	endpoint := &corev1.Endpoints{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      service,
+			Namespace: defaultNs,
+		},
+		Subsets: []corev1.EndpointSubset{
+			{
+				Addresses: []corev1.EndpointAddress{
+					{IP: "192.168.10.100"},
+				},
+			},
+		},
+	}
+	expectInit := []string{
+		"192.168.10.100:8080",
+		"192.168.10.100:9090",
+	}
+
+	cl := fake.NewSimpleClientset(endpoint)
+	res, err := newK8sResolver(cl, zap.NewNop(), service, ports)
+	require.NoError(t, err)
+
+	// step-1 initialize
+	require.NoError(t, res.start(context.Background()))
+	defer func() {
+		require.NoError(t, res.shutdown(context.Background()))
+	}()
+	_, err = res.resolve(context.Background()) // second resolution, should be noop
+
+	// verify endpoints should be the same as expectInit
+	assert.NoError(t, err)
+	assert.Equal(t, expectInit, res.endpoints)
+
+	// step-2 update, simulate changes to the backend ip address
+	exist := endpoint.DeepCopy()
+	endpoint.Subsets = append(endpoint.Subsets, corev1.EndpointSubset{
+		Addresses: []corev1.EndpointAddress{{IP: "10.10.0.11"}},
+	})
+	patch := client.MergeFrom(exist)
+	data, err := patch.Data(endpoint)
+	require.NoError(t, err)
+	_, err = cl.CoreV1().Endpoints(defaultNs).
+		Patch(context.TODO(), service, types.MergePatchType, data, metav1.PatchOptions{})
+	require.NoError(t, err)
+
+	// verify update
+	assert.Eventually(t, func() bool {
+		_, err := res.resolve(context.Background())
+		assert.NoError(t, err)
+		assert.Equal(t, []string{
+			"10.10.0.11:8080",
+			"10.10.0.11:9090",
+			"192.168.10.100:8080",
+			"192.168.10.100:9090",
+		}, res.endpoints)
+		return true
+	}, time.Second, 20*time.Millisecond)
+
+	// step-3 delete, simulate deletion of backends
```

Comment:
I'd prefer to have this on its own test.

Reply:
Sorry, I didn't get your point here, you mean a separate test method for the delete operation?

Reply:
Sorry, I meant that the step 3 should be on its own test function: instead of having one test doing everything, have smaller tests exercising smaller units of your code.